tensorflow_recommenders_addons/dynamic_embedding/python/keras/layers/embedding.py

```diff
@@ -0,0 +1,459 @@
+"""
+Dynamic Embedding is designed for Large-scale Sparse Weights Training.
+See [Sparse Domain Isolation](https://github.com/tensorflow/community/pull/237)
+"""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import tensorflow as tf
+
+from tensorflow.python.eager import context
+from tensorflow.python.ops import init_ops
+from tensorflow_recommenders_addons import dynamic_embedding as de
+from tensorflow_recommenders_addons.dynamic_embedding.python.ops import dynamic_embedding_variable as devar
+
+
+def _choose_reduce_method(combiner, sparse=False, segmented=False):
+  select = 'sparse' if sparse else 'math'
+  try:
+    module = getattr(tf, select)
+  except:
+    raise AttributeError('tensorflow has no attribute {}'.format(select))
+  select = 'segment_' if segmented else 'reduce_'
+  select += combiner
+  try:
+    method = getattr(module, select)
+  except:
+    raise AttributeError('Module [{}] has no attribute {}'.format(
+        module, select))
+  if not callable(method):
+    raise ValueError('{}: {} in {} is not callable'.format(
+        select, method, module))
+  return method
+
+
+@tf.function
+def reduce_pooling(x, combiner='sum'):
+  """
+  Default combine_fn for Embedding layer. By assuming input
+  ids shape is (batch_size, s1, ..., sn), it will get lookup result
+  with shape (batch_size, s1, ..., sn, embedding_size). Every
+  sample in a batch will be reduecd to a single vector, and thus
+  the output shape is (batch_size, embedding_size)
+  """
+
+  ndims = x.shape.ndims
+  combiner = _choose_reduce_method(combiner, sparse=False, segmented=False)
+
+  with tf.name_scope('deep_squash_pooling'):
+
+    if ndims == 1:
+      raise ValueError("reduce_pooling need at least dim-2 input.")
+    elif ndims == 2:
+      return combiner(x, 0)
+
+    for i in range(0, ndims - 2):
+      x = combiner(x, 1)
+    return x
+
+
+class Embedding(tf.keras.layers.Layer):
+  """
+  A keras style Embedding layer. It combines two parts of functionalities,
+  including lookup from the embedding parameter and get a numerical continuous
+  tensors to express the feature weights, refering to the discrete feature ids.
+
+  The embedding layer allow arbirary input shape of feature ids, and get
+  (shape(ids) + embedding_size) lookup result. Finally the output of the
+  layer with be squashed to (batch_size, embedding_size) if the input is
+  batched, or (embedding_size) if it's a single example.
+
+  You could inherit the `Embedding` class to implement a custom embedding
+  layer with other fixed shape output.
+
+  Please notice that the fixed shape output is required by the keras layer
+  mechanism. It you are working with a variant output situation, please
+  use basic APIs like `ShadowVariable` with AutoGraph/EagerMode or
+  `dynamic_embedding.embedding_lookup` in graph mode.
+  """
+
+  def __init__(self,
+               embedding_size,
+               combiner='sum',
```

Comment:
I recommend move down the `combiner` to the value dtype.

Reply:
Accept